File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-3.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */


func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode{
    dummy := &ListNode{}
    dummyHead := dummy

    for l1 != nil && l2 != nil{
        if l1.Val < l2.Val{
            dummy.Next = l1
            l1 = l1.Next
        }else{
            dummy.Next = l2
            l2 = l2.Next
        }
        dummy = dummy.Next
    }

    if l1 != nil{
        dummy.Next = l1
    }
    if l2 != nil{
        dummy.Next = l2
    }
    return dummyHead.Next
}

func mergeKLists(lists []*ListNode) *ListNode {
    if len(lists) == 0{
        return nil
    }
    for i := 1; i < len(lists); i++{
        lists[i] = mergeTwoLists(lists[i-1], lists[i])
    }

    return lists[len(lists)-1]
}

File: Data Structures & Algorithms/reorder-linked-list/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        # rev the second half
        slow = head
        fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        prev = None
        cur = slow.next
        slow.next = None
        while cur:
            next_ = cur.next
            cur.next = prev
            prev = cur
            cur = next_
        # the mid point may need to be pointed toward None before reconnecting links ...
        # use the ref to the og head and the rev head and reconnect links

        head2 = prev
        while head and head2:
            h1_next = head.next
            h2_next = head2.next
            head.next = head2
            head = h1_next
            head2.next = h1_next
            head2 = h2_next